File: apps/server/src/services/ChatService.ts
import { MettaParser, Post } from './MettaParser.js';
import { MorkService } from './MorkService.js';
import { XenovaEmbeddingService } from './EmbeddingService.js';

export class ChatService {
    private morkService: MorkService;
    private embeddingService: XenovaEmbeddingService;
    private mistralUrl: string;
    private mistralKey: string;

    constructor() {
        const morkUrl = process.env.MORK_URL || 'http://localhost:8001';
        this.mistralUrl = process.env.MISTRAL_API_URL || 'https://api.mistral.ai/v1';
        this.mistralKey = process.env.MISTRAL_API_KEY || '';

        this.morkService = new MorkService(morkUrl);
        this.embeddingService = new XenovaEmbeddingService();
    }

    // Ingestion
    async ingestData(userId: string, filePath?: string): Promise<string> {
        let resolvedPath = filePath;
        if (!resolvedPath) {
            const envPath = process.env.DATA_PATH;
            if (envPath) {
                // @ts-ignore
                const path = await import('path');
                resolvedPath = path.resolve(process.cwd(), envPath);
            } else {
                // @ts-ignore
                const path = await import('path');
                resolvedPath = path.resolve(__dirname, '../../../../data/data.metta');
            }
        }

        await this.morkService.clearGraph(userId);

        // @ts-ignore
        const posts = await MettaParser.parse(resolvedPath!, true, userId);

        // Upload to User Namespace
        for (const post of posts) {
            // Generate Embedding
            // @ts-ignore
            const textToEmbed = MettaParser.toEmbeddingString(post);
            const embedding = await this.embeddingService.embed(textToEmbed);

            // Upload Embedding to Namespace
            await this.morkService.uploadEmbeddings(post.id, embedding, userId);
            
        }
        
        return `Ingested ${posts.length} posts into namespace ${userId}.`;
    }

    // Retrieve
    async retrieve(userId: string, query: string, topK: number = 5): Promise<any[]> {
        // embed user query
        const queryEmbedding = await this.embeddingService.embed(query);

        // Fetch embeddings from MORK (User Namespace)
        const allEmbeddingsFlat = await this.morkService.getAllEmbeddings(userId);

        // Reconstruct vectors
        const vectors: Record<string, number[]> = {};

        // Added safe check in case vectors are empty
        if (allEmbeddingsFlat) {
            for (const { id, index, value } of allEmbeddingsFlat) {
                if (!vectors[id]) vectors[id] = [];
                vectors[id][index - 1] = value;
            }
        }

        // Cosine Similarity
        const scores = Object.entries(vectors).map(([id, vector]) => {
            for (let i = 0; i < queryEmbedding.length; i++) {
                if (vector[i] === undefined) vector[i] = 0;
            }

            const score = this.cosineSimilarity(queryEmbedding, vector);
            return { id, score };
        });

        // Top K
        scores.sort((a, b) => b.score - a.score);
        const top = scores.slice(0, topK);

        // Retrieve Content
        const results = [];
        for (const item of top) {
            const content = await this.morkService.getPostById(item.id, userId);
            results.push({ ...item, content });
        }

        return results;
    }

    private cosineSimilarity(a: number[], b: number[]): number {
        let dot = 0;
        let magA = 0;
        let magB = 0;
        for (let i = 0; i < a.length; i++) {
            dot += a[i] * b[i];
            magA += a[i] * a[i];
            magB += b[i] * b[i];
        }
        return dot / (Math.sqrt(magA) * Math.sqrt(magB) || 1);
    }

    // Chat
    // Mistral is used for chat generation(for free api), can switch to larger LLM for better result 
    async chat(message: string, historyContext: string = '', userId: string): Promise<string> {
        const context = await this.retrieve(userId, message);

        const contextStr = context.map(c => `[ID: ${c.id}]\n${c.content}`).join('\n---\n');

        try {
            const systemPrompt = `You are an intelligent assistant helping the user analyze their own posts and content. 
            The provided context contains specific posts with attributes like title, tone, complexity, and engagement. 
            Your goal is to answer briefly the user's question by generalizing patterns and inferring insights from this context. 
            Do not just list the retrieved items; instead, synthesize the information to provide a comprehensive and thoughtful answer based on the user's history. use the retrieved content to answer the question if only if the users question needs the context, otherwise answer the question without using the context. If the context is not relevant to the question, ignore it.`;

            let userPrompt = `Context from Knowledge Base:\n${contextStr}\n\n`;
            if (historyContext) {
                userPrompt += `Conversation History (relevant branch):\n${historyContext}\n\n`;
            }
            userPrompt += `Question: ${message}\nAnswer:`;

            const response = await fetch(`${this.mistralUrl}/chat/completions`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.mistralKey}`
                },
                body: JSON.stringify({
                    model: 'mistral-tiny',
                    messages: [
                        { role: 'system', content: systemPrompt },
                        { role: 'user', content: userPrompt }
                    ]
                })
            });

            if (!response.ok) {
                const err = await response.text();
                throw new Error(`Mistral Chat Error: ${err}`);
            }

            const data = await response.json();
            return data.choices[0].message.content;

        } catch (e: any) {
            throw new Error(`LLM Generation failed: ${e.message}`);
        }
    }
}

File: apps/server-py/services/chat_service.py
import os
import httpx
from dotenv import load_dotenv
from services import embedding_service, chromadb_service
from services.metta_parser import parse, to_embedding_string

load_dotenv()

MISTRAL_URL = os.getenv("MISTRAL_API_URL",
"https://api.mistral.ai/v1")
MISTRAL_KEY = os.getenv("MISTRAL_API_KEY", "")


async def ingest_data(user_id: str, file_path: str) -> str:
    posts = parse(file_path)
    chromadb_service.clear_collection(user_id)

    embeddings = [embedding_service.embed(to_embedding_string(p)) for p in posts]
    chromadb_service.add_documents(user_id, posts, embeddings)

    return f"Ingested {len(posts)} posts into namespace {user_id}."

async def retrieve(user_id: str, query: str, top_k: int = 5) -> dict:
    query_embedding = embedding_service.embed(query)
    results = chromadb_service.search(user_id, query_embedding, top_k)

    return {
        "results": results,
        "debug": {
            "query": query, 
            "top_documents": results.get("documents", [[]])[0]
        },
    }

async def chat(
    message: str,
    history_context: str,
    user_id: str,
    provider: str | None = None,
    api_key: str | None = None,
) -> dict:
    retrieval = await retrieve(user_id, message)
    documents = retrieval["results"].get("documents", [[]])[0]
    context_str = "\n---\n".join(documents)

    system_prompt = (
        "You are an intelligent assistant. The provided context contains specific posts. "
        "Synthesize the information to answer the user's question. "
        "Use context only if relevant; otherwise answer directly."
    )

    user_prompt = f"Context:\n{context_str}\n\n"

    if history_context:
        user_prompt += f"Conversation History:\n{history_context}\n\n"
    
    user_prompt += f"Question: {message}\nAnswer:"

    messages = [
        {"role":"system", "content":system_prompt},
        {"role":"user", "content":user_prompt}
    ]

    async with httpx.AsyncClient(timeout=200) as client:
        if provider == "openai":
            key = api_key or os.getenv("OPENAI_API_KEY", "")
            resp = await client.post(
                "https://api.openai.com/v1/chat/completions",
                headers={"Authorization":f"Bearer {key}"},
                json={"model":"gpt-3.5-turbo", "messages": messages},
            )

        else:
            key = api_key or MISTRAL_KEY
            resp = await client.post(
                f"{MISTRAL_URL}/chat/completions",
                headers={"Authorization": f"Bearer {key}"},
                json={"model":"mistral-tiny", "messages":messages},
            )
        
    resp.raise_for_status()
    text = resp.json()["choices"][0]["message"]["content"]
    return {"response": text, "debug": retrieval["debug"]}